README.md
# F0.Analyzers
Code refactorings, diagnostic suppressors and diagnostic analyzers with code fixes for C# projects, based on the .NET Compiler Platform ("Roslyn") Code Analysis.

![F0.Analyzers](https://raw.githubusercontent.com/Flash0ver/F0/master/Branding/GitHub/F0.Analyzers.png)

[![F0.Analyzers on fuget.org](https://www.fuget.org/packages/F0.Analyzers/badge.svg)](https://www.fuget.org/packages/F0.Analyzers)

## NuGet
[NuGet package](https://www.nuget.org/packages/F0.Analyzers/)

## Visual Studio Marketplace
[Visual Studio Extension](https://marketplace.visualstudio.com/items?itemName=Flash0Ware.F0-Analyzers-VS)

## API Browser / Package Viewer
[NuGet Package Explorer](https://nuget.info/packages/F0.Analyzers)\
[FuGet Package Explorer](https://www.fuget.org/packages/F0.Analyzers)\
[DotNetApis](http://dotnetapis.com/pkg/F0.Analyzers)\
[NuGet Must Haves](https://nugetmusthaves.com/Package/F0.Analyzers)\
[NuGet Trends](https://nugettrends.com/packages?months=12&ids=F0.Analyzers)

## Other external links
[Libraries.io Packages](https://libraries.io/nuget/F0.Analyzers)\
[Libraries.io Repositories](https://libraries.io/github/Flash0ver/F0.Analyzers)

## Documentation
See [F0.Analyzers.md](./documentation/F0.Analyzers.md).

## Changelog
See [CHANGELOG.md](./CHANGELOG.md).

## Contributing Guidelines
See [CONTRIBUTING.md](./CONTRIBUTING.md).

## Compatibility
requires _Roslyn v4.0.1_ or later (see [packages of the .NET Compiler Platform](https://github.com/dotnet/roslyn/blob/main/docs/wiki/NuGet-packages.md))
### .NET 6.0
supports [SDK 6.0.100](https://github.com/dotnet/core/blob/main/release-notes/6.0/6.0.0/6.0.0.md) and later
### Visual Studio 2022
supports [17.0.0](https://docs.microsoft.com/en-us/visualstudio/releases/2022/release-notes-v17.0) and later
### Rider
supports [2021.3](https://www.jetbrains.com/rider/whatsnew/2021-3/) and later
### Visual Studio Code
supports _OmniSharp (C#)_ [1.24.1](https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.24.1) and later


CHANGELOG.md
# F0.Analyzers
CHANGELOG

## vNext

## v0.10.0 (2022-03-21)
### Analyzers
- Added `IdentifiersShouldNotContainUnderscores` diagnostic suppressor, suppressing _Warning CA1707_ for _MSTest_, _NUnit_ and _xUnit.net_ test methods.

## v0.9.0 (2022-02-04)
### Analyzers
- Added `ImplicitRecordClassDeclaration` diagnostic analyzer, reporting _Warning F02001_ for `record` declarations without the `class` or `struct` keyword.
- Added `DeclareRecordClassExplicitly` code fix provider, adding the optional `class` keyword to `record` declarations that are _reference types_.
- Fixed `PreferPatternMatchingNullCheckOverComparisonWithNull` diagnostic analyzer, throwing `System.NullReferenceException` on invocations of `bool`-returning `static` methods named `Equals` with a single `null` argument.

### NuGet package
- Added support for `.NET SDK 6.0`.
- Removed support for `.NET SDK 5.0`.

### Visual Studio Extension
- Added support for `Visual Studio 2022`.
- Removed support for `Visual Studio 2017` and `Visual Studio 2019`.

## v0.8.0 (2021-10-13)
### Analyzers
- Changed `ObjectInitializer` code refactoring, additionally supporting _implicit object creation expressions_ (i.e. _target-typed `new` expressions_), introduced in _C# 9.0_.

## v0.7.1 (2021-06-13)
### Analyzers
- Fixed `UsePatternMatchingNullCheckInsteadOfComparisonWithNull` code fix provider, when fixing `null` check expressions in arguments.

## v0.7.0 (2021-05-11)
### Analyzers
- Added `PreferPatternMatchingNullCheckOverComparisonWithNull` diagnostic analyzer, reporting _Warning F01001_ for `null` check by equality, reporting _Info F01002_ for `null` check by identity.
- Added `UsePatternMatchingNullCheckInsteadOfComparisonWithNull` code fix provider, replacing both equality `null` checks (_F01001_) and identity `null` checks (_F01002_) with either `is null` or `is not null` pattern expressions.

### NuGet package
- Changed target framework from `.NET Standard 1.3` to `.NET Standard 2.0`.

## v0.6.1 (2021-04-18)
### Analyzers
- Fixed `ObjectInitializer` code refactoring, no longer registering code actions with collection types.

### Visual Studio Extension
- Removed _install_ and _uninstall_ scripts (NuGet tools).

## v0.6.0 (2020-11-21)
### Analyzers
- Changed `ObjectInitializer` code refactoring, additionally assigning inherited members from `base` classes.

## v0.5.0 (2020-09-29)
### Analyzers
- Changed `ObjectInitializer` code refactoring, additionally assigning `internal` fields and properties within files in the same assembly or _friend_ assemblies.

## v0.4.1 (2020-07-11)
### Analyzers
- Fixed `ObjectInitializer` code refactoring, assigning values to static members, causing compiler error CS1914.

## v0.4.0 (2020-06-26)
### Analyzers
- Changed `ObjectInitializer` code refactoring, assigning matching _locals_ and _members_ if existing in the current context.

### NuGet package
- Suppressed package dependencies.

## v0.3.0 (2020-05-27)
### Analyzers
- Added `GoToStatementConsideredHarmful` diagnostic analyzer, reporting _Warning F00001_ if a `goto` statement is used.

## v0.2.0 (2020-05-21)
### Analyzers
- Changed `ObjectInitializer` code refactoring, assigning `default operator` if language version is C# 7.0 or lower.
- Changed `ObjectInitializer` code refactoring, registering code action only if language version is C# 3.0 or greater.

### NuGet package
- Marked as a development-only dependency, which prevents the package from being included as a dependency in other packages.

## v0.1.0 (2020-05-14)
### Analyzers
- Added `ObjectInitializer` code refactoring, assigning `default literal` to all publicly settable fields and properties.

### NuGet package
- Uploaded first version, targeting `.NET Standard 1.3`.

### Visual Studio Extension
- Published initial version, supporting `Visual Studio 2017` and `Visual Studio 2019`.


scripts/Test-PackageLocally.ps1
#Requires -Version 6.0

[CmdletBinding()]
param (
    [Parameter(Mandatory=$false)]
    [Alias('c','configuration')]
    [ValidateSet('Debug', 'Release')]
    [string]$BuildConfiguration = 'Debug'
)

Set-StrictMode -Version 3.0
$ErrorActionPreference = 'Stop'

$RepositoryRootPath = (Get-Item -Path $PSScriptRoot).Parent
$ProjectName = 'F0.Analyzers'
$NuGetConfigurationFile = Join-Path -Path $RepositoryRootPath -ChildPath 'nuget.config'
$ProjectDirectory = Join-Path -Path $RepositoryRootPath -ChildPath 'source' -AdditionalChildPath 'package', 'F0.Analyzers.Package'
$ProjectFile = Join-Path -Path $ProjectDirectory -ChildPath 'F0.Analyzers.Package.csproj'
$PackageFile = Join-Path -Path $ProjectDirectory -ChildPath 'bin' -AdditionalChildPath $BuildConfiguration, '*.nupkg'
$PropertiesFile = Join-Path -Path $RepositoryRootPath -ChildPath 'source' -AdditionalChildPath 'Release.props'
$VersionXPath = '/Project/PropertyGroup/F0Version'
$VersionSuffix = 'local'
$LocalFeedName = 'local-feed'
$PackageSourcePath = "./$LocalFeedName"
$LocalFeedDirectory = Join-Path -Path $RepositoryRootPath -ChildPath $LocalFeedName
$ExampleDirectory = Join-Path -Path $RepositoryRootPath -ChildPath 'source' -AdditionalChildPath 'example'
$ExampleProjects =  Get-ChildItem -Path $ExampleDirectory -Filter '*.csproj' -Recurse -Depth 1 -File

function Add-LocalFeed {
    Write-Host 'create directory' -ForegroundColor Blue

    $output = New-Item -Path $RepositoryRootPath -Name $LocalFeedName -ItemType 'directory'

    Write-Host '     ' -NoNewline
    $output | Select-Object -Last 1 | Write-Host
}

function New-GitignoreFile {
    Write-Host 'create gitignore' -ForegroundColor Blue

    $GitignoreFileName = '.gitignore'
    $GitignoreFileContent = "*$([Environment]::NewLine)!$GitignoreFileName$([Environment]::NewLine)"
    $output = New-Item -Path $LocalFeedDirectory -Name $GitignoreFileName -ItemType 'file' -Value $GitignoreFileContent

    Write-Host '     ' -NoNewline
    $output | Select-Object -Last 1 | Write-Host
}

function Remove-LocalFeed {
    Write-Host 'delete directory' -ForegroundColor Blue

    Remove-Item -Path $LocalFeedDirectory -Recurse

    Write-Host "     $LocalFeedDirectory"
}

function Add-PackageSource {
    Write-Host 'dotnet nuget add source' -ForegroundColor Blue

    $output = dotnet nuget add source $PackageSourcePath --configfile $NuGetConfigurationFile --name $LocalFeedName

    Write-Host "     $output"
}

function Remove-PackageSource {
    Write-Host 'dotnet nuget remove source' -ForegroundColor Blue

    $output = dotnet nuget remove source $LocalFeedName --configfile $NuGetConfigurationFile

    Write-Host "     $output"
}

function Clean-Output {
    Write-Host 'dotnet clean' -ForegroundColor Blue

    $output = dotnet clean $ProjectFile --configuration $BuildConfiguration

    $output | Select-Object -Last 7 | Select-Object -First 1 | Write-Host
}

function Build-Package {
    Write-Host 'dotnet pack' -ForegroundColor Blue

    $output = dotnet pack $ProjectFile --configuration $BuildConfiguration --version-suffix $VersionSuffix

    Write-Host '   ' -NoNewline
    $output | Select-Object -Last 1 | Write-Host
}

function Deploy-PackageLocally {
    Write-Host 'dotnet nuget push' -ForegroundColor Blue

    $output = dotnet nuget push $PackageFile --source $LocalFeedDirectory

    Write-Host '     ' -NoNewline
    $output | Select-Object -Last 1 | Write-Host
}

function Install-Package {
    Write-Host 'dotnet add package' -ForegroundColor Blue

    $VersionPrefix = Select-Xml -Path $PropertiesFile -XPath $VersionXPath | Select-Object -ExpandProperty Node | Select-Object -ExpandProperty InnerText
    $Version = "$VersionPrefix-$VersionSuffix"

    foreach ($ExampleProject in $ExampleProjects) {
        $output = dotnet add $ExampleProject package $ProjectName --package-directory $LocalFeedDirectory --source $LocalFeedDirectory --version $Version

        Write-Host '   - ' -NoNewline
        $output | Select-Object -Last 4 | Select-Object -First 1 | Write-Host
    }
}

function Uninstall-Package {
    Write-Host 'dotnet remove package' -ForegroundColor Blue

    foreach ($ExampleProject in $ExampleProjects) {
        $output = dotnet remove $ExampleProject package $ProjectName

        Write-Host '   - ' -NoNewline
        Write-Host $output
    }
}

function Wait-ManualTest {
    Write-Host 'Setup completed successfully. You may now test the Analyzers locally via the Examples. When finished, press Enter/Return to continue with Teardown . . .' -ForegroundColor Yellow

    $Input = Read-Host
}

function Write-TestInvoke {
    Write-Host 'Manual local test invoked.' -ForegroundColor Magenta
}

function Write-TestComplete {
    Write-Host 'Manual local test completed.' -ForegroundColor Green
}

function Write-Version {
    $output = dotnet --version

    Write-Host "     .NET SDK $output"
}

Write-TestInvoke
Write-Version
Clean-Output
Build-Package
Add-LocalFeed
New-GitignoreFile
Deploy-PackageLocally
Add-PackageSource
Install-Package
Wait-ManualTest
Uninstall-Package
Remove-PackageSource
Remove-LocalFeed
Write-TestComplete
